pkg/kubelet/dynamicpodspec/ports.go

```diff
@@ -0,0 +1,155 @@
+package dynamicpodspec
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	netutil "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/klog"
+
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+)
+
+type assignPortAdmitHandler struct {
+	podAnnotation string
+	portRange     netutil.PortRange
+	podUpdater    PodUpdater
+}
+
+func NewAssignPortHandler(podAnnotation string, portRange netutil.PortRange, podUpdater PodUpdater) *assignPortAdmitHandler {
+	return &assignPortAdmitHandler{
+		podAnnotation: podAnnotation,
+		portRange:     portRange,
+		podUpdater:    podUpdater,
+	}
+}
+
+func (w *assignPortAdmitHandler) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	pod := attrs.Pod
+	_, autoport := pod.ObjectMeta.Annotations[w.podAnnotation]
+	if !autoport {
+		return lifecycle.PodAdmitResult{
+			Admit: true,
+		}
+	}
+
+	usedPorts, lastUsedPort := getUsedPorts(attrs.OtherPods...)
+	count := 0
+	for i := range pod.Spec.Containers {
+		for j := range pod.Spec.Containers[i].Ports {
+			if pod.Spec.Containers[i].Ports[j].HostPort == 0 {
+				count++
+			}
+		}
+	}
+
+	if count == 0 {
+		return lifecycle.PodAdmitResult{
+			Admit: true,
+		}
+	}
+
+	port, isAssigned := getAvaPort(usedPorts, w.portRange.Base, w.portRange.Size,
+		len(usedPorts), lastUsedPort, count)
+	if isAssigned {
+		for i := range pod.Spec.Containers {
+			for j := range pod.Spec.Containers[i].Ports {
+				if pod.Spec.Containers[i].Ports[j].HostPort != 0 {
+					continue
+				}
+				pod.Spec.Containers[i].Ports[j].HostPort = int32(port)
+				if pod.Spec.HostNetwork {
+					pod.Spec.Containers[i].Ports[j].ContainerPort = int32(port)
+				}
+				envVariable := v1.EnvVar{
+					Name:  fmt.Sprintf("PORT%d", j),
+					Value: fmt.Sprintf("%d", pod.Spec.Containers[i].Ports[j].HostPort),
+				}
+				pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, envVariable)
+				if j == 0 {
```

Comment:
It seems no necessary to set this env.

Reply:
for bussiness process to fetch assigned port?

Reply:
it's already set in `PORT0`, set `PORT` env is a compatibility code。

Reply:
ok